src/core/solvers/box-qp.cpp

```diff
@@ -0,0 +1,258 @@
+    alphas_[n] = 1. / pow(2., static_cast<double>(n));
+  }
+}
+
+BoxQP::~BoxQP() {}
+
+const BoxQPSolution& BoxQP::solve(const Eigen::MatrixXd& H, const Eigen::VectorXd& q, const Eigen::VectorXd& lb,
+                                  const Eigen::VectorXd& ub, const Eigen::VectorXd& xinit) {
+  if (static_cast<std::size_t>(H.rows()) != nx_ || static_cast<std::size_t>(H.cols()) != nx_) {
+    throw_pretty("Invalid argument: "
+                 << "H has wrong dimension (it should be " + std::to_string(nx_) + "," + std::to_string(nx_) + ")");
+  }
+  if (static_cast<std::size_t>(q.size()) != nx_) {
+    throw_pretty("Invalid argument: "
+                 << "q has wrong dimension (it should be " + std::to_string(nx_) + ")");
+  }
+  if (static_cast<std::size_t>(lb.size()) != nx_) {
+    throw_pretty("Invalid argument: "
+                 << "lb has wrong dimension (it should be " + std::to_string(nx_) + ")");
+  }
+  if (static_cast<std::size_t>(ub.size()) != nx_) {
+    throw_pretty("Invalid argument: "
+                 << "ub has wrong dimension (it should be " + std::to_string(nx_) + ")");
+  }
+  if (static_cast<std::size_t>(xinit.size()) != nx_) {
+    throw_pretty("Invalid argument: "
+                 << "xinit has wrong dimension (it should be " + std::to_string(nx_) + ")");
+  }
+
+  // We need to enforce feasible warm-starting of the algorithm
+  for (std::size_t i = 0; i < nx_; ++i) {
+    x_(i) = std::max(std::min(xinit(i), ub(i)), lb(i));
+  }
+
+  // Start the numerical iterations
+  for (std::size_t k = 0; k < maxiter_; ++k) {
+    clamped_idx_.clear();
+    free_idx_.clear();
+    // Compute the gradient
+    g_ = q;
+    g_.noalias() += H * x_;
+    for (std::size_t j = 0; j < nx_; ++j) {
+      const double& gj = g_(j);
+      const double& xj = x_(j);
+      const double& lbj = lb(j);
+      const double& ubj = ub(j);
+      if ((xj == lbj && gj > 0.) || (xj == ubj && gj < 0.)) {
+        clamped_idx_.push_back(j);
+      } else {
+        free_idx_.push_back(j);
+      }
+    }
+
+    // Check convergence
+    nf_ = free_idx_.size();
+    nc_ = clamped_idx_.size();
+    if (g_.lpNorm<Eigen::Infinity>() <= th_grad_ || nf_ == 0) {
+      if (k == 0) {  // compute the inverse of the free Hessian
+        Hff_.resize(nf_, nf_);
+        for (std::size_t i = 0; i < nf_; ++i) {
+          const std::size_t& fi = free_idx_[i];
+          for (std::size_t j = 0; j < nf_; ++j) {
+            Hff_(i, j) = H(fi, free_idx_[j]);
+          }
+        }
+        if (reg_ != 0.) {
+          Hff_.diagonal().array() += reg_;
+        }
+        Hff_inv_llt_ = Eigen::LLT<Eigen::MatrixXd>(nf_);
+        Hff_inv_llt_.compute(Hff_);
+        Eigen::ComputationInfo info = Hff_inv_llt_.info();
+        if (info != Eigen::Success) {
+          throw_pretty("backward_error");
+        }
+        Hff_inv_.setIdentity(nf_, nf_);
+        Hff_inv_llt_.solveInPlace(Hff_inv_);
+      }
+      solution_.Hff_inv = Hff_inv_;
```

Comment:
You can avoid one copy here, by working directly with `solution_.Hff_inv`

Reply:
done. I also avoid copy of the free and clamped indexes